python3Packages.ruff: reuse already built ruff binary

bengsparks opened the pull request:
The maturin build system cannot see that we already build the ruff binary in nixpkgs. This leads to the ruff binary being rebuilt and placed inside of the resulting Python package. Every nixpkgs-review that passes through the `pkgs.ruff` package therefore builds `ruff` 3 times per architecture. Said binary was never even used due to the patching that already occurs in `find_ruff_bin`.

Especially with how time-consuming release builds can be in Rust, there is no reason to keep this package in this state.

master: `nix-build -A python3Packages.ruff` - `0.44s user 0.28s system 0% cpu 5:09.64 total`
branch: `nix-build -A python3Packages.ruff` - `0.36s user 0.12s system 9% cpu 4.856 total`

~~Still requires a symlink at the appropriate place that points at the ruff binary; I'll get to it later today.~~

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

GaetanLepage reviewed (commented):
This is a great idea! It could be done for `uv` too.
Is the PR ready?

bengsparks committed:
python3Packages.ruff: reuse already built ruff binary

The maturin build system cannot see that we already build the ruff binary in nixpkgs.
This leads to the ruff binary being rebuilt and placed inside of the resulting Python package.
Every nixpkgs-review that passes through the `pkgs.ruff` package therefore builds `ruff` 3 times per architecture.
Said binary was never even used due to the patching that already occurs in `find_ruff_bin`.

Especially with how time-consuming release builds can be in Rust, there is no reason to keep this package in this state.

master: `nix-build -A python3Packages.ruff` - `0.44s user 0.28s system 0% cpu 5:09.64 total`
branch: `nix-build -A python3Packages.ruff` - `0.36s user 0.12s system 9% cpu 4.856 total`

bengsparks commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 412112`

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 80 packages built:</summary>
  <ul>
    <li>dmarc-metrics-exporter</li>
    <li>dmarc-metrics-exporter.dist</li>
    <li>litellm</li>
    <li>litellm.dist</li>
    <li>marimo (python312Packages.marimo)</li>
    <li>marimo.dist (python312Packages.marimo.dist)</li>
    <li>openapi-python-client</li>
    <li>openapi-python-client.dist</li>
    <li>oterm</li>
    <li>oterm.dist</li>
    <li>python312Packages.fastapi-mcp</li>
    <li>python312Packages.fastapi-mcp.dist</li>
    <li>python312Packages.fastmcp</li>
    <li>python312Packages.fastmcp.dist</li>
    <li>python312Packages.gradio</li>
    <li>python312Packages.gradio.dist</li>
    <li>python312Packages.mcp</li>
    <li>python312Packages.mcp.dist</li>
    <li>python312Packages.mcpadapt</li>
    <li>python312Packages.mcpadapt.dist</li>
    <li>python312Packages.mypy-boto3-builder</li>
    <li>python312Packages.mypy-boto3-builder.dist</li>
    <li>python312Packages.pyhanko</li>
    <li>python312Packages.pyhanko.dist</li>
    <li>python312Packages.pytest-examples</li>
    <li>python312Packages.pytest-examples.dist</li>
    <li>python312Packages.pytest-ruff</li>
    <li>python312Packages.pytest-ruff.dist</li>
    <li>python312Packages.python-lsp-ruff</li>
    <li>python312Packages.python-lsp-ruff.dist</li>
    <li>python312Packages.reflex</li>
    <li>python312Packages.reflex-chakra</li>
    <li>python312Packages.reflex-chakra.dist</li>
    <li>python312Packages.reflex.dist</li>
    <li>python312Packages.ruff</li>
    <li>python312Packages.ruff.dist</li>
    <li>python312Packages.smolagents</li>
    <li>python312Packages.smolagents.dist</li>
    <li>python312Packages.xhtml2pdf</li>
    <li>python312Packages.xhtml2pdf.dist</li>
    <li>python312Packages.xsdata</li>
    <li>python312Packages.xsdata.dist</li>
    <li>python313Packages.fastapi-mcp</li>
    <li>python313Packages.fastapi-mcp.dist</li>
    <li>python313Packages.fastmcp</li>
    <li>python313Packages.fastmcp.dist</li>
    <li>python313Packages.gradio</li>
    <li>python313Packages.gradio.dist</li>
    <li>python313Packages.marimo</li>
    <li>python313Packages.marimo.dist</li>
    <li>python313Packages.mcp</li>
    <li>python313Packages.mcp.dist</li>
    <li>python313Packages.mcpadapt</li>
    <li>python313Packages.mcpadapt.dist</li>
    <li>python313Packages.mypy-boto3-builder</li>
    <li>python313Packages.mypy-boto3-builder.dist</li>
    <li>python313Packages.pyhanko</li>
    <li>python313Packages.pyhanko.dist</li>
    <li>python313Packages.pytest-examples</li>
    <li>python313Packages.pytest-examples.dist</li>
    <li>python313Packages.pytest-ruff</li>
    <li>python313Packages.pytest-ruff.dist</li>
    <li>python313Packages.python-lsp-ruff</li>
    <li>python313Packages.python-lsp-ruff.dist</li>
    <li>python313Packages.reflex</li>
    <li>python313Packages.reflex-chakra</li>
    <li>python313Packages.reflex-chakra.dist</li>
    <li>python313Packages.reflex.dist</li>
    <li>python313Packages.ruff</li>
    <li>python313Packages.ruff.dist</li>
    <li>python313Packages.smolagents</li>
    <li>python313Packages.smolagents.dist</li>
    <li>python313Packages.xhtml2pdf</li>
    <li>python313Packages.xhtml2pdf.dist</li>
    <li>python313Packages.xsdata</li>
    <li>python313Packages.xsdata.dist</li>
    <li>vectorcode</li>
    <li>vectorcode.dist</li>
    <li>vunnel</li>
    <li>vunnel.dist</li>
  </ul>
</details>

Prince213 reviewed (approved):
Built on x86_64-linux and build logs LGTM

GaetanLepage reviewed (approved):
LGTM, thanks!
